### PR TITLE
types: make channel types a lot stricter

### PR DESCRIPTION
**Pull request by kyranet**
**Please describe the changes this PR makes and why it should be merged:**

Brings a new set of strict channel types based on @ImRodry's typeguards from #6957. This change also makes sure that the set of possible channel types are strict, explicit, and matching the actual classes obtained during runtime.

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Review by vladfrangu — APPROVED** [review]

File: typings/index.d.ts [hunk 1]
```diff
@@ -521,9 +521,9 @@ export abstract class Channel extends Base {
   public id: Snowflake;
   public readonly partial: false;
   public type: keyof typeof ChannelTypes;
-  public delete(): Promise<Channel>;
-  public fetch(force?: boolean): Promise<Channel>;
-  public isText(): this is TextBasedChannels;
+  public delete(): Promise<this>;
+  public fetch(force?: boolean): Promise<this>;
```
Comment (ImRodry) [inline typings/index.d.ts:525]: Didn’t @suneettipirneni say something about returning `this` and how it didn’t make sense in this situation?

Comment (suneettipirneni) [reply typings/index.d.ts:525]: It makes sense here, because if you narrow a type the promise on this class has the narrowed return type as well.

Comment (ImRodry) [reply typings/index.d.ts:525]: Oh alright, when does it not make sense then?

Comment (suneettipirneni) [reply typings/index.d.ts:525]: The other PR was more of an aesthetic thing, the type that was discussed before wasn't narrowed. But even in the case before it's fine.

Comment (ImRodry) [reply typings/index.d.ts:525]: Alright, thanks

**Review by iCrawl — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**